### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Importing dependencies required
import pandas as pd
import os
import csv
import numpy as np

# File to Load (Remember to Change These)

file = os.path.join("Resources","purchase_data.csv")

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file)

# Printing the head to sample the data
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [2]:
# Printing the number of unique players

purchase_data["SN"].nunique()

576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
# Getting the number of unique items into a variable:
unq_items=purchase_data["Item ID"].nunique()

# Getting the mean of price into a variable:
avg_price = purchase_data["Price"].mean()

# Getting the count  of purchase_id column into a variable:
ttl_purchases = purchase_data["Purchase ID"].count()

# Getting the number of sum of Price as revenue into a variable:
ttl_revenue = purchase_data["Price"].sum()

# Building the DF with the above generated variables
summary1_df = pd.DataFrame({"Number of Unique Items":[unq_items],
                            "Average Price":[avg_price],
                            "Number of Purchases":[ttl_purchases],
                            "Total Revenue":[ttl_revenue]
                            })

# Formating the Columns
summary1_df["Average Price"] = summary1_df["Average Price"].map("${:.2f}".format)
summary1_df["Total Revenue"] = summary1_df["Total Revenue"].map("${:.2f}".format)

# Printing the DF
summary1_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780      $2379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed


In [4]:
# Grouping by Gender and getting count of unique screen names and renaming it to Total count
GD_total_count = purchase_data.groupby(["Gender"]).nunique()["SN"].rename("Total Count")
#print(GD_total_count)

# Storing the total number of unique players into a variable
ttl_players=purchase_data["SN"].nunique()

#Storing the total count data frame into a temp DF
temp = GD_total_count

#Building the percentage of players DF
POP=(temp/ttl_players)*100
#print(POP)

# Combining the two DFs to form the final DF as expected result
Gender_Demographics = pd.DataFrame({"Total Count":GD_total_count,
                                    "Percentage of Players":POP
                                  })

# Formating column values of DF
Gender_Demographics["Percentage of Players"]=Gender_Demographics["Percentage of Players"].map("{:.2f}%".format)

#printing DF
Gender_Demographics


Total Count Percentage of Players
Gender                                                  
Female                          81                14.06%
Male                           484                84.03%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
# Copying the original DF to retain the original
purchaseAnaGen = purchase_data.copy()

# Grouping by Gender to get count and renaming it as Purchase Count(colum name)
purchaseAnaGen_count = purchaseAnaGen.groupby("Gender").count()["SN"].rename("Purchase Count")
#print(purchaseAnaGen_count)

# Grouping by Gender to get mean of price and renaming it as Avg Purchase Price(colum name)
purchaseAnaGen_mean = purchaseAnaGen.groupby("Gender").mean()["Price"].rename("Average Purchase Price")
#print(purchaseAnaGen_mean)

# Grouping by Gender to get sum of price and renaming it as Total Purchase value(colum name)
purchaseAnaGen_sum = purchaseAnaGen.groupby("Gender").sum()["Price"].rename("Total Purchase Value")
#print(purchaseAnaGen_sum)

# Grouping by Gender to get count of unique SN's
purchaseAnaGen_unique_count = purchaseAnaGen.groupby("Gender").nunique("SN")["SN"]
#print(purchaseAnaGen_unique_count)

# Building the Average total purchase per person
purchaseAnaGen_AvgperPerson=purchaseAnaGen_sum/purchaseAnaGen_unique_count
#print(purchaseAnaGen_AvgperPerson)

# Building the resultant DF from above individual DFs
df_PurchaseAnalysisByGender = pd.DataFrame({"Purchase Count":purchaseAnaGen_count,
                                            "Average Purchase Price":purchaseAnaGen_mean,
                                            "Total Purchase Value":purchaseAnaGen_sum,
                                            "Avg Total Purchase per Person":purchaseAnaGen_AvgperPerson
                                            })
# Formating the column values
df_PurchaseAnalysisByGender["Average Purchase Price"] = df_PurchaseAnalysisByGender["Average Purchase Price"].map("${:.2f}".format)
df_PurchaseAnalysisByGender["Total Purchase Value"] = df_PurchaseAnalysisByGender["Total Purchase Value"].map("${:.2f}".format)
df_PurchaseAnalysisByGender["Avg Total Purchase per Person"] = df_PurchaseAnalysisByGender["Avg Total Purchase per Person"].map("${:.2f}".format)

# Displaying the result
df_PurchaseAnalysisByGender


Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
# Defining the BINs and Groups
bins = [0,9,14,19,24,29,34,39,100]
group=["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

# Copying the original DF to retain original DF for future use
reorgd_AgeAna = purchase_data.copy()

# Dropping duplicate name rows
reorgd_AgeAna = reorgd_AgeAna.drop_duplicates(subset="SN",keep='first',inplace=False)

# Binning and labeling by group values
reorgd_AgeAna["Age Group"] = pd.cut(reorgd_AgeAna["Age"],bins,labels=group)

# Calculating total rows
total= reorgd_AgeAna["SN"].count()
#print(total)

# Retaining only the required two columns for analysis
reorgd_AgeAna = reorgd_AgeAna[["Age", "Age Group"]]

# Grouping by Age Group column
gb_ageGrp = reorgd_AgeAna.groupby(["Age Group"])

# Counting the values by Age
new_df = gb_ageGrp.count()

# Renaming the column Age as Total Count
renmd_df = new_df.rename(columns={"Age":"Total Count"})

# Calculating percentage of players in each age group
percentage_of_Players=(renmd_df["Total Count"]/total)*100

# Adding the Percentage of players column to the renamed DF built before
renmd_df["Percentage of Players"]= percentage_of_Players

#Formating the DF
renmd_df["Percentage of Players"]=renmd_df["Percentage of Players"].map("{:.2f}%".format)

#Printing the DF
renmd_df


Total Count Percentage of Players
Age Group                                   
<10                 17                 2.95%
10-14               22                 3.82%
15-19              107                18.58%
20-24              258                44.79%
25-29               77                13.37%
30-34               52                 9.03%
35-39               31                 5.38%
40+                 12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
# Defining the BINs and groups
bins = [0,9,14,19,24,29,34,39,100]

group=["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

#Creating a copy of original data frame
reorgd_PurAgeAna = purchase_data.copy()

# Binning
reorgd_PurAgeAna["Age Ranges"] = pd.cut(reorgd_PurAgeAna["Age"],bins,labels=group)

# Building two DFs with two columns required for later calculation - 1. count 2. sum/mean
new_PurAgeAna_Count = reorgd_PurAgeAna[["Age", "Age Ranges"]]
new_PurAgeAna_SumAvg = reorgd_PurAgeAna[["Price", "Age Ranges"]]

# Building the temp DFs for count, mean and sum
temp1 = new_PurAgeAna_Count.groupby(["Age Ranges"]).count()["Age"].rename("Purchase Count")
temp2= new_PurAgeAna_SumAvg.groupby(["Age Ranges"]).mean()["Price"].rename("Avg Purchase Price")
temp3=new_PurAgeAna_SumAvg.groupby(["Age Ranges"]).sum()["Price"].rename("Total Purchase Value")

# Saving the total count of unique SNs into a variable
ageGrp_count = renmd_df["Total Count"]

# Building the DF with Count, Mean and Sum DFs
df_purchaseAnalysisAge = pd.DataFrame({"Purchase Count":temp1,
                           "Avg Purchase Price":temp2,
                           "Total Purchase Value":temp3,
                            "Avg Total Purchase per Person":temp3/ageGrp_count
                          })

# Formating columns
df_purchaseAnalysisAge["Avg Purchase Price"]=df_purchaseAnalysisAge["Avg Purchase Price"].map("${:.2f}".format)
df_purchaseAnalysisAge["Total Purchase Value"]=df_purchaseAnalysisAge["Total Purchase Value"].map("${:.2f}".format)
df_purchaseAnalysisAge["Avg Total Purchase per Person"]=df_purchaseAnalysisAge["Avg Total Purchase per Person"].map("${:.2f}".format)

# Printing the DF
df_purchaseAnalysisAge

Purchase Count Avg Purchase Price Total Purchase Value  \
Age Ranges                                                           
<10                     23              $3.35               $77.13   
10-14                   28              $2.96               $82.78   
15-19                  136              $3.04              $412.89   
20-24                  365              $3.05             $1114.06   
25-29                  101              $2.90              $293.00   
30-34                   73              $2.93              $214.00   
35-39                   41              $3.60              $147.67   
40+                     13              $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [8]:
# Creating a copy of the original data
topspenders = purchase_data.copy()

# Grouping by ScreenName
gb_topspenders = topspenders.groupby("SN")

# Calculating and storing the count of Purchase ID
gb_topspenders_count = gb_topspenders.count()["Purchase ID"]

# Calculating and storing the mean of Price
gb_topspenders_mean = gb_topspenders.mean()["Price"]

# Calculating and storing the total price
gb_topspenders_sum = gb_topspenders.sum()["Price"]

# Combining the new DataFrame with the Count, Mean & Sum
df_topspenders = pd.DataFrame({ "Purchase Count":gb_topspenders_count,
                                "Average Purchase Price": gb_topspenders_mean,
                                "Total Purchase Value": gb_topspenders_sum
                              })

#df_topspenders.sort_values(by="Total Purchase Value",ascending=False)

# Sorting, displaying the head and formating of column values as needed
df_topspenders=df_topspenders.sort_values(by="Total Purchase Value",ascending=False).head().style.format({"Average Purchase Price":"${:.2f}","Total Purchase Value":"${:.2f}"})
#df_topspenders["Total Purchase Value"]=df_topspenders["Total Purchase Value"].style.format("${:.2f}")

# Printing the DF
df_topspenders




## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [16]:
# Making a copy of original dataframe with required columns 
MPI_df = purchase_data[["Item ID","Item Name", "Price"]]
#MPI_df

# Grouping by two columns - Item ID and Item Name
gb_MPI_df = MPI_df.groupby(["Item ID","Item Name"])

# Calculating count & sum and saving it
gb_MPI_df_count = gb_MPI_df.count()["Price"]
gb_MPI_df_sum = gb_MPI_df.sum()["Price"]

# Building the new DF
df_MPI = pd.DataFrame({"Purchase Count": gb_MPI_df_count,
                       "Item Price":gb_MPI_df_sum/gb_MPI_df_count,
                       "Total Purchase Value":gb_MPI_df_sum
                      })

# Sorting, Print head and formating of the dataframe
df_MPI.sort_values(by="Purchase Count",ascending=False).head().style.format({"Item Price":"${:.2f}","Total Purchase Value":"${:.2f}"})



## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [20]:
# Changing the sort by value to 'Total Purchase Value'
df_MPI.sort_values(by="Total Purchase Value",ascending=False).head().style.format({"Item Price":"${:.2f}","Total Purchase Value":"${:.2f}"})

## Three observable trends:

1. Close to 84% of gamers playing Heroes of Pymoli are males

2. Gamers in the 20-24 age group are the biggest spenders with 45% of total spending

3. Item 'Oathbreaker, Last Hope of the Breaking Storm' has sold the most and is also the most profitable item